### PR TITLE
Cut generated app url hash suffix to 8 characters

### PR DESCRIPTION
**Pull request by StrongMonkey**
The current generated url uses the first 12 characters of a hash of serviceName + appName + namespace, which is a bit longer and undesirable. Cut down to 8 to make url a little bit nicer.

This change will probably need to be made into changelog for the next release as it changes the format of user's app url.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ var (
 	DefaultImageCheckIntervalDefault = "5m"
 
 	// Default HttpEndpointPattern set to enable Let's Encrypt
-	DefaultHttpEndpointPattern = "{{printf \"%s-%s-%s\" .Container .App .Hash | truncate}}.{{.ClusterDomain}}"
+	DefaultHttpEndpointPattern = "{{getAppFullName .Container .App .Namespace 8 | truncate}}.{{.ClusterDomain}}"
```
Comment (StrongMonkey) [inline pkg/config/config.go:35]: Made a new function to make it easiler to pass parameter and limit of hash

File: pkg/publish/ingress.go [hunk 2]
```diff
@@ -104,13 +104,19 @@ func toEndpoint(pattern, domain, container, appName, appNamespace string) (strin
 	return templateBuffer.String(), nil
 }
 
-func getAppHash(serviceName, appName, appNamespace string) string {
+/*
+getAppFullName takes containerName, appName and namespace, convert it to the format of
+`${containerName}-${appName}-${hash of `containerName:appName:namespace`}, with
+the ability to specify the length of the hash suffix
+*/
+func getAppFullName(serviceName, appName, appNamespace string, limit int) string {
```
Comment (StrongMonkey) [inline pkg/publish/ingress.go:112]: made it a go template function so that user can override the limit as they want.

**Comment by StrongMonkey** [comment]
This will reconsile all existing workloads to short name if users don't put a default value.

It will break if user uses cname record to point the old DNS name. 

To revert the behavior, pass `{{getAppFullName .Container .App .Namespace 12 | truncate}}.{{.ClusterDomain}}"` to `--http-endpoint-pattern`

**Review by ibuildthecloud — APPROVED** [review]

File: pkg/publish/ingress.go [hunk 3]
```diff
@@ -78,13 +78,13 @@ func toEndpoint(pattern, domain, container, appName, appNamespace string) (strin
 		App:           appName,
 		Container:     container,
 		Namespace:     appNamespace,
-		Hash:          getAppHash(container, appName, appNamespace),
```
Comment (ibuildthecloud) [inline pkg/publish/ingress.go:81]: Can you keep this field here and have it call hashConcat with size 8.  Otherwise there's no way to update production which is using the `Hash` field in the template pattern.

**Review by ibuildthecloud — CHANGES_REQUESTED** [review]

File: pkg/publish/ingress.go (hunk 3 above)
Comment (StrongMonkey) [reply pkg/publish/ingress.go:81]: Oh shoot, I will just keep it

Comment (StrongMonkey) [reply pkg/publish/ingress.go:81]: Ok, I added this back.

**Review by tylerslaton — COMMENTED** [review]
Changes look good! Would you mind adding test cases to `ingress_test.go` that accounts for the various states of `hashConcat`? For example - what happens if we give it something other than 8 or some invalid input.

**Review by ibuildthecloud — DISMISSED** [review]
Can you address the comment to add a unit test for edge case of hashConcat

**Comment by StrongMonkey** [comment]
@tylerslaton unit tests are added 👍

**Review by tylerslaton — APPROVED** [review]
Won't block as this is a minor thing, but what happens if I set something like:

```
{{hashConcat 8 .Foo .Bar .Baz | truncate}}-{{.Namespace}}.{{.ClusterDomain}}
```

Would be good to see that in a test as a well so we know the UX of erroneous templates passed is consistent.

**Comment by StrongMonkey** [comment]
@tylerslaton Good point! Added

**Review by cjellick — APPROVED** [review]
